core/trino-spi/src/main/java/io/trino/spi/predicate/Utils.java

```diff
@@ -64,4 +72,47 @@ static RuntimeException handleThrowable(Throwable throwable)
         }
         return new RuntimeException(throwable);
     }
+
+    /**
+     * Try to expand {@code valueSet} into a discrete set of (at most {@code limit}) objects.
+     * For example: [1, 5] can be expanded into {1, 2, 3, 4, 5}.
+     * If the data type is not supported or the expansion results in too many values, {@code Optional.empty()} is returned.
+     */
+    public static Optional<Collection<Object>> tryExpandRanges(ValueSet valuesSet, int limit)
+    {
+        if (!(valuesSet instanceof SortedRangeSet) || limit <= 0) {
+            return Optional.empty();
+        }
+        List<Range> ranges = valuesSet.getRanges().getOrderedRanges();
+        Type type = valuesSet.getType();
+        List<Object> result = new ArrayList<>(limit);
+        for (Range range : ranges) {
+            if (range.isSingleValue()) {
+                result.add(range.getSingleValue());
+                continue;
+            }
+            if (range.isLowUnbounded() || range.isHighUnbounded()) {
+                return Optional.empty();
+            }
+            Optional<Stream<?>> discreteValues = type.getDiscreteValues(new Type.Range(range.getLowBoundedValue(), range.getHighBoundedValue()));
+            if (discreteValues.isEmpty()) {
+                return Optional.empty();
+            }
+            Iterator<?> iterator = discreteValues.get().iterator();
+            if (!range.isLowInclusive()) {
+                iterator.next();
```

Comment:
add 
`verify( hasNext(), ...)`

Reply:
Is it OK to use `com.google.common.base.Verify.verify` in SPI?
I wasn't sure so I added an explicit `if-then-throw` check.